### PR TITLE
implemented redux

### PR DESCRIPTION
**Pull request by maxniceman**

### CODE REVIEW

File: src/hooks/useActions.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import { useDispatch } from "react-redux";
+import { actions as cartActions } from "../store/cartSlice";
+import { actions as favoriteActions } from "../store/favoritesSlice";
+
+const rootActions = {
+  ...cartActions,
+  ...favoriteActions,
+};
+
+export const useActions = () => {
```
Comment (ValeriyaZavozova) [inline src/hooks/useActions.ts:12]: I would rather separate actions by store slices. For instance `useCartActions` for `cart`

Comment (ValeriyaZavozova) [inline src/hooks/useActions.ts]: Or better, you can create `useCartStore` hook which returns all actions and store from `useSelector((state: RootState) => state.cart)`

File: src/pages/ProductPage/ProductPage.tsx [hunk 2]
```diff
@@ -20,9 +23,12 @@ export const loader: LoaderFunction = async ({ params }) => {
 
 const ProductPage = () => {
   const { id, image, title, price, description } = useLoaderData() as Product;
+  const { favorites } = useSelector((state: RootState) => state.favorites);
+  const isFavorite = favorites.some((f) => f.id === id);
```
Comment (ValeriyaZavozova) [inline src/pages/ProductPage/ProductPage.tsx]: Better to create a separate method in `useFavoritesStore`.
```
const useFavoritesStore = () => {
  ....
  const { favorites } = useSelector((state: RootState) => state.favorites);

  const checkIfItemAdded = useCallback((id) => {
    return favorites.some((f) => f.id === id);
  }, [])

  ...
  
  return { ..., favorites, checkIfItemAdded }
```

File: src/store/cartSlice.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+};
+
+export const cartSlice = createSlice({
+  name: "cart",
+  initialState,
+  reducers: {
+    add: (state, action: PayloadAction<CartProduct>) => {
+      const index = state.cart.findIndex((c) => c.id === action.payload.id);
+      if (index !== -1) return;
+      else state.cart.push(action.payload);
```
Comment (ValeriyaZavozova) [inline src/store/cartSlice.ts]:
```suggestion
      const isAlreadyInCart = state.cart.some((c) => c.id === action.payload.id);
      if (isAlreadyInCart) return;
      else state.cart.push(action.payload);
```

File: src/store/cartSlice.ts [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    },
+    remove: (state, action: PayloadAction<{ id: number }>) => {
+      const index = state.cart.findIndex((c) => c.id === action.payload.id);
+      state.cart.splice(index, 1);
```
Comment (ValeriyaZavozova) [inline src/store/cartSlice.ts]: What will be if index is `-1` (item is not in cart)

File: src/store/favoritesSlice.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+};
+
+export const favoritesSlice = createSlice({
+  name: "favorites",
+  initialState,
+  reducers: {
+    toggleFavorites: (state, action: PayloadAction<FavoriteProduct>) => {
+      const index = state.favorites.findIndex(
+        (f) => f.id === action.payload.id
+      );
```
Comment (ValeriyaZavozova) [inline src/store/favoritesSlice.ts]: use `some` and more descriptive variable name

Comment (maxniceman) [reply src/store/favoritesSlice.ts]: renamed variable but didn't use some method because I need the index of the element to remove it then

**Review by ValeriyaZavozova — APPROVED** [review]

**Pull request merged**